### PR TITLE
Add migration to fix signature keys inconsistency

### PR DESCRIPTION
**Pull request by mshaposhnik**
### What does this PR do?

Adds migration that makes sure there is all workspace public & private pairs exists, check that all workspaces have unique key pairs, and adds appropriate constraints to the DB schema.

### What issues does this PR fix or reference?
#11087 

#### Release Notes
N/A

#### Docs PR
N/A

### CODE REVIEW

File: wsmaster/che-core-sql-schema/src/main/resources/che-schema/6.11.0/1__add_signature_key_constraints.sql [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+--
+-- SPDX-License-Identifier: EPL-2.0
+--
+-- Contributors:
+--   Red Hat, Inc. - initial API and implementation
+--
+
+
+-- remove key pair records which linked to non-existing keys
+DELETE FROM che_sign_key_pair kp
```
Comment (sleshchenko) [inline wsmaster/che-core-sql-schema/src/main/resources/che-schema/6.11.0/1__add_signature_key_constraints.sql:15]: Can't it be simplified like the following?
```
DELETE FROM che_sign_key_pair kp
WHERE NOT EXISTS (
   SELECT id
   FROM   che_sign_key k
   WHERE  k.id = kp.private_key OR k.id = kp.public_key
   )
```

File: wsmaster/che-core-sql-schema/src/main/resources/che-schema/6.11.0/1__add_signature_key_constraints.sql [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+      WHERE kp1.private_key  = kp2.private_key
+      OR kp1.private_key = kp2.public_key
+      ) > 1;
+
+
+-- add pair uniqueness constraint
+CREATE UNIQUE INDEX index_sign_public_private_key_id ON che_sign_key_pair (public_key, private_key);
+
+-- add keys table constraints
+ALTER TABLE che_sign_key_pair ADD CONSTRAINT fk_sign_public_key_id FOREIGN KEY (public_key) REFERENCES che_sign_key (id);
```
Comment (sleshchenko) [inline wsmaster/che-core-sql-schema/src/main/resources/che-schema/6.11.0/1__add_signature_key_constraints.sql]: Make sure that there are indexes for `fk_sign_public_key_id` and `fk_sign_private_key_id`

File: wsmaster/che-core-sql-schema/src/main/resources/che-schema/6.11.0/1__add_signature_key_constraints.sql [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+   FROM   che_sign_key k
+   WHERE  k.id = kp.public_key
+   );
+
+-- remove duplicated keys id if any
+DELETE FROM che_sign_key_pair kp1
+WHERE (
+      SELECT count(*) FROM che_sign_key_pair kp2
+      WHERE kp1.private_key  = kp2.private_key
+      OR kp1.private_key = kp2.public_key
```
Comment (sleshchenko) [inline wsmaster/che-core-sql-schema/src/main/resources/che-schema/6.11.0/1__add_signature_key_constraints.sql]: I guess you wanted to write `kp1.public_key = kp2.public_key` here

**Review by sleshchenko — COMMENTED** [review]
Take a look my inlined comments

File: wsmaster/che-core-sql-schema/src/main/resources/che-schema/6.11.0/1__add_signature_key_constraints.sql (hunk 1 above)
Comment (mshaposhnik) [reply wsmaster/che-core-sql-schema/src/main/resources/che-schema/6.11.0/1__add_signature_key_constraints.sql:15]: Nope.  That query will work only if both private_  and public_ key not exists.

Comment (sleshchenko) [reply wsmaster/che-core-sql-schema/src/main/resources/che-schema/6.11.0/1__add_signature_key_constraints.sql:15]: I understand. Thanks.

There is a possible case when key par A exists, and it has public key but not private one. Key pair is going to be removed. But private key won't be removed. As result adding foreign key will fail. Am I right? If yes, please consider adding cleaning private and public keys which are referencing non-existing key pair.

Comment (sleshchenko) [reply wsmaster/che-core-sql-schema/src/main/resources/che-schema/6.11.0/1__add_signature_key_constraints.sql]: I understand. Thanks.

There is a possible case when key par A exists, and it has public key but not private one. Key pair is going to be removed. But private key won't be removed. As result adding foreign key will fail. Am I right? If yes, please consider adding cleaning private and public keys which are referencing non-existing key pair.

File: wsmaster/che-core-sql-schema/src/main/resources/che-schema/6.11.0/1__add_signature_key_constraints.sql (hunk 2 above)
Comment (mshaposhnik) [reply wsmaster/che-core-sql-schema/src/main/resources/che-schema/6.11.0/1__add_signature_key_constraints.sql]: Didn't get the idea. Explain, pls.

Comment (sleshchenko) [reply wsmaster/che-core-sql-schema/src/main/resources/che-schema/6.11.0/1__add_signature_key_constraints.sql]: >  We never do seatch on that fields

I would say that we don't do such searches explicitly. But I think JPA uses such search to fetch keys for a requested pair.

We have to add indexes on foreign keys, and we do it like here https://github.com/eclipse/che/blob/6c53555fded7621014501ab44d2c4bbca869b4db/wsmaster/che-core-sql-schema/src/main/resources/che-schema/5.8.0/1__add_foreigh_key_indexes.sql#L18-L17

Correct me If I'm wrong.

Comment (mshaposhnik) [reply wsmaster/che-core-sql-schema/src/main/resources/che-schema/6.11.0/1__add_signature_key_constraints.sql]: Ok, i didn't remember the reason, but seems it's needed for  JPA. Will add separate indexes.

File: wsmaster/che-core-sql-schema/src/main/resources/che-schema/6.11.0/1__add_signature_key_constraints.sql (hunk 1 above)
Comment (sleshchenko) [reply wsmaster/che-core-sql-schema/src/main/resources/che-schema/6.11.0/1__add_signature_key_constraints.sql:15]: > I don't see reason why foreign key shoud fail?

You're right. I missed that foreign keys are from pair to keys but not wise versa.

Comment (mshaposhnik) [reply wsmaster/che-core-sql-schema/src/main/resources/che-schema/6.11.0/1__add_signature_key_constraints.sql:15]: Added extra unused keys cleanup query

Comment (mshaposhnik) [reply wsmaster/che-core-sql-schema/src/main/resources/che-schema/6.11.0/1__add_signature_key_constraints.sql:15]: https://github.com/eclipse/che/pull/11113/files#diff-f71f338e81773023418fd98cf10db36aR36

**Comment by skabashnyuk** [comment]
ci-test

**Comment by riuvshin** [comment]
ci-test build report:
[Build details](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/756/)
[Test report](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/756//Selenium_20tests_20report/)
[selenium tests report data](http://tests.codenvycorp.com/selenium_test_report_data/11113)
docker image: eclipseche/che-server:11113
https://github.com/orgs/eclipse/teams/eclipse-che-qa please check this report.

**Comment by skabashnyuk** [comment]
ci-test

**Comment by riuvshin** [comment]
ci-test build report:
[Build details](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/757/)
[Test report](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/757//Selenium_20tests_20report/)
[selenium tests report data](http://tests.codenvycorp.com/selenium_test_report_data/11113)
docker image: eclipseche/che-server:11113
https://github.com/orgs/eclipse/teams/eclipse-che-qa please check this report.

**Review by skabashnyuk — APPROVED** [review]

**Review by sleshchenko — APPROVED** [review]

**Pull request merged**